lib/unparser/preprocessor.rb

```diff
@@ -151,7 +151,53 @@ def chunked_children
 
     end # CollapseStrChildren
 
-    # Preprocessor eliminating unneded dstr nodes
+    # Preprocessor flattening unneeded dstr nesting. This happens when a dstr is generated from
+    # implicit string concatenation, and one of the strings being concatenated is itself a dstr.
+    #
+    # Example: "foo" "bar #{baz}"
+    #
+    # Without this preprocessor, this turns into: "foo#{"bar#{baz}"}"
+    # With this preprocessor, this turns into: "foobar #{baz}"
+    class FlattenImplicitDSTR < self
+
+      # No need to register :dsym, because Ruby doesn't do implicit concatenation of symbols
+      register :dstr
+
+      FLATTEN_CHILDREN = IceNine.deep_freeze([:dstr, :str])
```

Comment:
Not yet, the rest of the code base does not use 2.1 syntax yet. I want to do an atomic switchover. This gem just dropped support for < 2.1. The cleanup will be a followup to reduce the diff and keep it focused.